src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs

```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+    {
+        var isCollection = propType.CollectionKind != CodeTypeBase.CodeTypeCollectionKind.None;
+        if (propType.TypeDefinition is CodeEnum currentEnum)
+            return $"writeEnumValue<{currentEnum.Name.ToFirstCharacterUpperCase()}>";
+        else if (isCollection)
+        {
+            if (propType.TypeDefinition == null)
+                return $"writeCollectionOfPrimitiveValues<{propertyType.ToFirstCharacterLowerCase()}>";
+            else
+                return $"writeCollectionOfObjectValues";
+        }
+        return null;
+    }
+
+    private void WriteDeserializerFunction(CodeFunction codeFunction, LanguageWriter writer)
+    {
+        var param = codeFunction.OriginalLocalMethod.Parameters.FirstOrDefault();
+
+        if (param?.Type is CodeType codeType && codeType.TypeDefinition is CodeInterface codeInterface)
+        {
+
+            var inherits = codeInterface?.StartBlock?.Implements.FirstOrDefault(x => x.TypeDefinition is CodeInterface);
+
+            var properties = codeInterface?.Properties.Where(static x => x.Kind == CodePropertyKind.Custom && !x.ExistsInBaseType);
 
-        writer.WriteLine($"return new {returnType.ToFirstCharacterUpperCase()}();");
+            writer.WriteLine("return {");
+            writer.IncreaseIndent();
+            if (inherits != null)
+            {
+                writer.WriteLine($"...deserializeInto{inherits?.TypeDefinition?.Name.ToFirstCharacterUpperCase()}({param.Name.ToFirstCharacterLowerCase()}),");
+            }
+
+            if (properties != null)
+            {
+                foreach (var otherProp in properties)
+                {
+                    var keyName = !string.IsNullOrWhiteSpace(otherProp.SerializationName) ? otherProp.SerializationName.ToFirstCharacterLowerCase() : otherProp.Name.ToFirstCharacterLowerCase();
+                    writer.WriteLine($"\"{keyName}\": n => {{ {param.Name.ToFirstCharacterLowerCase()}.{otherProp.Name.ToFirstCharacterLowerCase()} = n.{GetDeserializationMethodName(otherProp.Type, codeFunction)}; }},");
+                }
+            }
+            writer.DecreaseIndent();
+            writer.WriteLine("}");
+        }
+        else
+            throw new InvalidOperationException($"Model interface for deserializer function {codeFunction.Name} is not available");
+    }
+
+    private string GetDeserializationMethodName(CodeTypeBase propType, CodeFunction codeFunction)
+    {
+        var isCollection = propType.CollectionKind != CodeTypeBase.CodeTypeCollectionKind.None;
+        var propertyType = localConventions?.GetTypeString(propType, codeFunction, false);
+        if (propertyType != null && propType is CodeType currentType)
+        {
+            if (currentType.TypeDefinition is CodeEnum currentEnum)
+                return $"getEnumValue{(currentEnum.Flags || isCollection ? "s" : string.Empty)}<{currentEnum.Name.ToFirstCharacterUpperCase()}>({propertyType.ToFirstCharacterUpperCase()})";
```

Comment:
isn't the collection method getCollectionOfEnumValues? instead of just getEnumValues?

Reply:
Itis getEnumValues according to:

https://github.com/microsoft/kiota/blob/main/src/Kiota.Builder/Writers/TypeScript/CodeMethodWriter.cs
https://github.com/microsoft/kiota-typescript/blob/main/packages/abstractions/src/serialization/parseNode.ts